src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java

```diff
@@ -405,13 +404,32 @@ class ModuleExtensionProxy implements Structure {
     private final Location location;
     private final HashBiMap<String, String> imports;
     private final ImmutableList.Builder<Tag> tags;
+    private final boolean devDependency;
 
     ModuleExtensionProxy(String extensionBzlFile, String extensionName, Location location) {
       this.extensionBzlFile = extensionBzlFile;
       this.extensionName = extensionName;
       this.location = location;
       this.imports = HashBiMap.create();
       this.tags = ImmutableList.builder();
+      this.devDependency = false;
+    }
+
+    private ModuleExtensionProxy(ModuleExtensionProxy other, boolean devDependency) {
+      this.extensionBzlFile = other.extensionBzlFile;
+      this.extensionName = other.extensionName;
+      this.location = other.location;
+      this.imports = other.imports;
+      this.tags = other.tags;
+      this.devDependency = devDependency;
+    }
+
+    /**
+     * Creates a proxy with the specified devDependency bit that shares accumulated imports and tags
+     * with the current one, thus preserving tag specification order.
+     */
+    ModuleExtensionProxy withDevDependency(boolean devDependency) {
+      return devDependency ? new ModuleExtensionProxy(this, true) : this;
```

Comment:
this might be a bit error-prone (it assumes `this` always has `devDependency=false`. Maybe just always create a new proxy object to reduce the mental burden? (proxy-proxy-proxies should work just fine, right...?)